Sources/SwiftParser/Diagnostics/ParseDiagnosticsGenerator.swift

```diff
@@ -131,7 +127,44 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
     if shouldSkip(node) {
       return .skipChildren
     }
-    addDiagnostic(node, position: node.endPosition, MissingNodeError(missingNode: Syntax(node)))
+
+    let position: AbsolutePosition
+    let fixIts: [FixIt]
+    if let token = Syntax(node).as(TokenSyntax.self) {
+      position = token.position
+      fixIts = [
+        FixIt(message: InsertTokenFixIt(missingToken: token), changes: [
+          .makePresent(node: node)
+        ])
+      ]
+    } else {
+      position = node.endPosition
+      fixIts = []
+    }
+
+    // Find the previous sibling, skipping over siblings that don't actually contain any source text.
```

Comment:
> Can't we just use `.sourceAccurate` here rather than checking for missing?

No, because then we wouldn’t find missing tokens, which are skipped in the `.sourceAccurate` traversal mode.

> Rather than grabbing previous siblings and replacing previous diagnostics, could we instead grab next siblings and just create the one diagnostic at this point? That would avoid looking through all diagnostics each time. IMO we also want to merge the fix-its, which that change would also make easier. 

Good idea. I thought that this would be more difficult than looking behind but it turns out that it’s actually simpler.

> We may want to start splitting these files up as well. Ie. the generation + message for this one could be defined in a `MissingSyntaxDiagnostic.swift` instead (it's a fair amount of code).

Good idea.